tiledb/sm/kv/kv.cc

```diff
@@ -64,92 +64,93 @@ uint64_t KV::capacity() const {
+Status KV::is_dirty(bool* dirty) const {
+  if (query_type_ != QueryType::WRITE)
+    return LOG_STATUS(Status::KVError(
+        "Cannot check if dirty; Key-value store was not opened in write mode"));
+
+  *dirty = !items_.empty();
+
+  return Status::Ok();
 }
 
-Status KV::init(const URI& kv_uri, const std::vector<std::string>& attributes) {
-  kv_uri_ = kv_uri;
+Status KV::open(QueryType query_type) {
+  if (is_open_)
+    RETURN_NOT_OK(Status::KVError(
+        "Cannot open key-value store; Key-value store already open"));
+
+  query_type_ = query_type;
 
   // Open the array for reads
-  if (open_array_for_reads_ == nullptr)
+  if (query_type_ == QueryType::READ) {
     RETURN_NOT_OK(storage_manager_->array_open(
-        kv_uri_, QueryType::READ, &open_array_for_reads_, &snapshot_));
+        kv_uri_, QueryType::READ, &open_array_, &snapshot_));
+  }
 
   // Open the array for writes
   uint64_t snapshot;  // will be ignored for writes
-  if (open_array_for_writes_ == nullptr)
+  if (query_type_ == QueryType::WRITE) {
     RETURN_NOT_OK(storage_manager_->array_open(
-        kv_uri_, QueryType::WRITE, &open_array_for_writes_, &snapshot));
-
-  schema_ = open_array_for_reads_->array_schema();
-
-  // Get attributes
-  if (attributes.empty()) {  // Load all attributes
-    write_good_ = true;
-    auto schema_attributes = schema_->attributes();
-    attributes_.push_back(constants::coords);
-    for (const auto& attr : schema_attributes)
-      attributes_.push_back(attr->name());
-  } else {  // Load input attributes
-    write_good_ = false;
-    attributes_.push_back(constants::coords);
-    attributes_.push_back(constants::key_attr_name);
-    attributes_.push_back(constants::key_type_attr_name);
-    for (const auto& attr : attributes) {
-      if (attr == constants::coords || attr == constants::key_type_attr_name ||
-          attr == constants::key_attr_name)
-        continue;
-      attributes_.push_back(attr);
-    }
+        kv_uri_, QueryType::WRITE, &open_array_, &snapshot));
   }
 
+  // Get array schema
+  schema_ = open_array_->array_schema();
+
+  // Load all attributes
+  auto schema_attributes = schema_->attributes();
+  attributes_.push_back(constants::coords);
+  for (const auto& attr : schema_attributes)
+    attributes_.push_back(attr->name());
+
   // Prepare attribute types and read buffer sizes
   for (const auto& attr : attributes_) {
     attribute_types_.push_back(schema_->type(attr));
     read_buffer_sizes_[attr] = std::pair<uint64_t, uint64_t>(0, 0);
   }
 
+  is_open_ = true;
+
   return Status::Ok();
 }
 
 void KV::clear() {
   schema_ = nullptr;
-  kv_uri_ = URI("");
   attributes_.clear();
   attribute_types_.clear();
 
   clear_items();
   clear_read_buffers();
   clear_write_buffers();
 
-  open_array_for_reads_ = nullptr;
-  open_array_for_writes_ = nullptr;
+  open_array_ = nullptr;
 }
 
-Status KV::finalize() {
-  RETURN_NOT_OK(flush());
+Status KV::close() {
+  if (!is_open_)
```

Comment:
Don't we return an error status in this case for regular arrays?

Reply:
No, we changed that recently to a noop: https://github.com/TileDB-Inc/TileDB/blob/dev/tiledb/sm/c_api/tiledb.cc#L2475.